File: Find Largest Value in Each Tree Row.java
// Time Complexity :O(N)
// Space Complexity :O(W), W- maximum width of the tree
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
       

        List<Integer> result = new ArrayList<>();
        if(root==null)
            return result;

         Queue<TreeNode> q= new LinkedList<>();
         q.add(root);

         while(!q.isEmpty())
         {
             int size =q.size();
             int maxi=Integer.MIN_VALUE;
             for(int i=0;i<size;i++)
             {
                 TreeNode curr= q.poll();

                 if(curr.left!=null)
                    q.add(curr.left);

                 if(curr.right!=null)
                    q.add(curr.right);

                    maxi= Math.max(maxi,curr.val);
             }
             result.add(maxi);
         }
       
        return result;
        
    }
}

File: findLargestValueInEachTreeRow.cpp
//time complexity:O(n)
//space complexity:O(n)
//executed on leetcode: yes
//approach:using bfs
//any issues faced? no

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        vector<int>res;
        if(root==NULL)
            return res;
        
        queue<TreeNode*>q;
        if(root)
            q.push(root);
        while(!q.empty())
        {
            int max=INT_MIN;
            int size=q.size();
            for(int i=0;i<size;i++)
            {
                TreeNode*curr=q.front();
                q.pop();
                if(curr->val>=max)
                    max=curr->val;
                if(curr->left)
                    q.push(curr->left);
                if(curr->right)
                    q.push(curr->right);
            }
            res.push_back(max);
        }
        return res;
    }
};